src/python/pants/backend/go/goals/generate.py

```diff
@@ -0,0 +1,285 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+import os
+import re
+import shlex
+import string
+from dataclasses import dataclass
+from typing import Mapping
+
+from pants.backend.go.target_types import GoPackageSourcesField
+from pants.backend.go.util_rules import first_party_pkg, goroot, sdk
+from pants.backend.go.util_rules.first_party_pkg import (
+    FallibleFirstPartyPkgAnalysis,
+    FallibleFirstPartyPkgDigest,
+    FirstPartyPkgAnalysis,
+    FirstPartyPkgAnalysisRequest,
+    FirstPartyPkgDigestRequest,
+)
+from pants.backend.go.util_rules.goroot import GoRoot
+from pants.build_graph.address import Address
+from pants.engine.env_vars import EnvironmentVars, EnvironmentVarsRequest
+from pants.engine.fs import DigestContents, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.native_engine import EMPTY_DIGEST, AddPrefix, Digest, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import collect_rules, goal_rule, rule, rule_helper
+from pants.engine.target import Targets
+from pants.option.option_types import StrListOption
+from pants.util.strutil import softwrap
+
+logger = logging.getLogger(__name__)
+
+
+# Adapted from Go toolchain.
+# See https://github.com/golang/go/blob/master/src/cmd/go/internal/generate/generate.go and
+# https://github.com/golang/go/blob/cc1b20e8adf83865a1dbffa259c7a04ef0699b43/src/os/env.go#L16-L96
+#
+# Original copyright:
+#   // Copyright 2011 The Go Authors. All rights reserved.
+#   // Use of this source code is governed by a BSD-style
+#   // license that can be found in the LICENSE file.
+
+
+_GENERATE_DIRECTIVE_RE = re.compile(rb"^//go:generate[ \t](.*)$")
+
+
+class GoGenerateGoalSubsystem(GoalSubsystem):
+    name = "go-generate"
+    help = softwrap(
+        """
+        Run each command in a package described by a `//go:generate` directive. This is equivalent to running
+        `go generate` on a Go package.
+
+        Note: Just like with `go generate`, the `go-generate` goal is never run as part of the build and
+        must be run manually to invoke the commands described by the //go:generate directives.
```

Comment:
My understanding is @ryanking would like it to be automated? Although, without having better Go experience, I generally bias towards your more conservative approach Tom.

Reply:
I think its ok to be conservative and not automate, as long as a user like myself could automate it. I guess I would do that either via a plugin or macro.

Reply:
Yeah, I think that you could wire this up into a codegen plugin. At that point, maybe we have an upstreamed option that determines whether the codegen rule should be used or not? That would be cool - you can always manually generate, or have it automated.

Although, it's a little duplicative to have the goal registered still because you could manually generate via `export-codegen`

Reply:
For the official Pants releases, I prefer we follow the direction of the underlying ecosystem / tool. For `go generate`, it is explicitly documented as never running automatically during a build. A regular user of the Go toolchain would likely be surprised if Pants did run generators during a build.

That said, the `RunPackageGeneratorsRequest` -> `RunPackageGeneratorsResult` should be easily invoked from a plugin for users that want custom behavior.